File: Classes/InfoWindowController.m
#import "InfoWindowController.h"
@import CoreServices.CarbonCore.Resources;
#import "ResourceDocument.h"
#import "Resource.h"
#import "ApplicationDelegate.h"
#import "NSOutlineView-SelectedItems.h"

@implementation InfoWindowController

- (void)windowDidLoad
{
	[super windowDidLoad];
	
	// set window to only accept key when editing text boxes
	[(NSPanel *)[self window] setBecomesKeyOnlyIfNeeded:YES];
	
    [self setMainWindow:NSApp.mainWindow];
	[self updateInfoWindow];
	
    [NSNotificationCenter.defaultCenter addObserver:self selector:@selector(mainWindowChanged:) name:NSWindowDidBecomeMainNotification object:nil];
	[NSNotificationCenter.defaultCenter addObserver:self selector:@selector(selectedResourceChanged:) name:NSOutlineViewSelectionDidChangeNotification object:nil];
	[NSNotificationCenter.defaultCenter addObserver:self selector:@selector(propertiesDidChange:) name:ResourceDidChangeNotification object:nil];
	[NSNotificationCenter.defaultCenter addObserver:self selector:@selector(propertiesDidChange:) name:DocumentInfoDidChangeNotification object:nil];
}

/*!
@method		updateInfoWindow
@updated	2003-11-06 NGS:	Fixed creator/type handling.
@updated	2003-10-26 NGS:	Now asks app delegate for icon instead of NSWorkspace.
@updated	2003-10-26 NGS:	Improved document name & icon display.
*/

- (void)updateInfoWindow
{
    nameView.editable = selectedResource != nil;
    nameView.bezeled = selectedResource != nil;
	
	if (selectedResource) {
		// set UI values
        self.window.title = NSLocalizedString(@"Resource Info", nil);
        nameView.stringValue = selectedResource.name;
        iconView.image = [(ApplicationDelegate *)[NSApp delegate] iconForResourceType:selectedResource.type];
        
		[[attributesMatrix cellWithTag:resChanged]		setState:selectedResource.attributes & resChanged];
		[[attributesMatrix cellWithTag:resPreload]		setState:selectedResource.attributes & resPreload];
		[[attributesMatrix cellWithTag:resProtected]	setState:selectedResource.attributes & resProtected];
		[[attributesMatrix cellWithTag:resLocked]		setState:selectedResource.attributes & resLocked];
		[[attributesMatrix cellWithTag:resPurgeable]	setState:selectedResource.attributes & resPurgeable];
		[[attributesMatrix cellWithTag:resSysHeap]	    setState:selectedResource.attributes & resSysHeap];

        rType.stringValue = GetNSStringFromOSType(selectedResource.type);
        rID.intValue = selectedResource.resID;
        rSize.integerValue = selectedResource.data.length;
        
        // swap box
        placeholderView.contentView = resourceView;
	} else if (currentDocument) {
		// get sizes of forks as they are on disk
		NSInteger dataLogicalSize = 0, rsrcLogicalSize = 0;
		
		// set info window elements to correct values
        self.window.title = NSLocalizedString(@"Document Info", nil);
		if (currentDocument.fileURL) {
            iconView.image = [[NSWorkspace sharedWorkspace] iconForFile:currentDocument.fileURL.path];
            nameView.stringValue = currentDocument.fileURL.lastPathComponent;
            NSNumber *dataSize, *totalSize;
            [currentDocument.fileURL getResourceValue:&dataSize forKey:NSURLFileSizeKey error:nil];
            [currentDocument.fileURL getResourceValue:&totalSize forKey:NSURLTotalFileSizeKey error:nil];
            dataLogicalSize = dataSize.integerValue;
            rsrcLogicalSize = totalSize.integerValue - dataLogicalSize;
		} else {
            iconView.image = [NSImage imageNamed:@"Resource file"];
            nameView.stringValue = currentDocument.displayName;
		}
		
        creator.stringValue = GetNSStringFromOSType(currentDocument.creator);
        type.stringValue = GetNSStringFromOSType(currentDocument.type);
        dataSize.integerValue = dataLogicalSize;
		rsrcSize.integerValue = rsrcLogicalSize;
		
		// swap box
        placeholderView.contentView = documentView;
	} else {
        iconView.image = nil;
        nameView.stringValue = @"";
        placeholderView.contentView = nil;
	}
}

- (void)setMainWindow:(NSWindow *)mainWindow
{
    NSWindowController *controller = mainWindow.windowController;
    if ([controller.document isKindOfClass:ResourceDocument.class]) {
        currentDocument = controller.document;
        [self setSelectedResource:currentDocument.outlineView.selectedItem];
    } else {
        currentDocument = nil;
        [self setSelectedResource:controller.resource];
    }
}

- (void)setSelectedResource:(Resource *)resource
{
    if ([resource isKindOfClass:Resource.class]) {
        selectedResource = resource;
    } else {
        selectedResource = nil;
    }
    [self updateInfoWindow];
}

- (void)mainWindowChanged:(NSNotification *)notification
{
    [self setMainWindow:notification.object];
}

- (void)selectedResourceChanged:(NSNotification *)notification
{
    NSOutlineView *outlineView = notification.object;
    if (outlineView.window.windowController.document == currentDocument) {
        [self setSelectedResource:[notification.object selectedItem]];
    }
}

- (void)propertiesDidChange:(NSNotification *)notification
{
	[self updateInfoWindow];
}

- (IBAction)creatorChanged:(id)sender
{
    currentDocument.creator = GetOSTypeFromNSString(creator.stringValue);
}

- (IBAction)typeChanged:(id)sender
{
    currentDocument.type = GetOSTypeFromNSString(type.stringValue);
}

- (IBAction)nameChanged:(id)sender
{
    selectedResource.name = nameView.stringValue;
}

- (IBAction)rTypeChanged:(id)sender
{
    selectedResource.type = GetOSTypeFromNSString(rType.stringValue);
    rType.stringValue = GetNSStringFromOSType(selectedResource.type); // Reload in case change was rejected
}

- (IBAction)rIDChanged:(id)sender
{
    selectedResource.resID = (ResID)rID.intValue;
    rID.intValue = selectedResource.resID;
}

- (IBAction)attributesChanged:(id)sender
{
    selectedResource.attributes ^= (short)[sender selectedTag];
}

+ (id)sharedInfoWindowController
{
	static InfoWindowController *sharedInfoWindowController = nil;
	if (!sharedInfoWindowController)
		sharedInfoWindowController = [[InfoWindowController allocWithZone:nil] initWithWindowNibName:@"InfoWindow"];
	return sharedInfoWindowController;
}

@end

@implementation NSWindowController (InfoWindowAdditions)

- (Resource *)resource
{
	return nil;
}

@end

File: Classes/InfoWindowController.swift
import Cocoa
import RKSupport

class InfoWindowController: NSWindowController, NSTextFieldDelegate {
    @IBOutlet var iconView: NSImageView!
    @IBOutlet var nameView: NSTextField!
    @IBOutlet var placeholderView: NSBox!
    @IBOutlet var resourceView: NSBox!
    @IBOutlet var documentView: NSBox!
    
    @IBOutlet var creator: NSTextField!
    @IBOutlet var type: NSTextField!
    @IBOutlet var dataSize: NSTextField!
    @IBOutlet var rsrcSize: NSTextField!
    
    @IBOutlet var rType: NSTextField!
    @IBOutlet var rID: NSTextField!
    @IBOutlet var rSize: NSTextField!
    @IBOutlet var attributesMatrix: NSMatrix!
    
    private var currentDocument: ResourceDocument!
    private var selectedResource: Resource!
    
    static var shared = InfoWindowController(windowNibName: "InfoWindow")
    
    override func windowDidLoad() {
        super.windowDidLoad()
        
        self.setMainWindow(NSApp.mainWindow)
        self.update()
        
        NotificationCenter.default.addObserver(self, selector: #selector(mainWindowChanged(_:)), name: NSWindow.didBecomeMainNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(selectedResourceChanged(_:)), name: NSOutlineView.selectionDidChangeNotification, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(propertiesChanged(_:)), name: .ResourceDidChange, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(propertiesChanged(_:)), name: .DocumentInfoDidChange, object: nil)
    }
    
    private func update() {
        nameView.abortEditing()
        nameView.isEditable = selectedResource != nil
        nameView.isBezeled = selectedResource != nil
        
        if let resource = selectedResource {
            // set UI values
            self.window?.title = NSLocalizedString("Resource Info", comment: "")
            nameView.stringValue = resource.name
            iconView.image = ApplicationDelegate.icon(for: resource.type)
            
            attributesMatrix.cell(withTag: ResAttributes.changed.rawValue)?.state    = resource.attributes.contains(.changed) ? .on : .off
            attributesMatrix.cell(withTag: ResAttributes.preload.rawValue)?.state    = resource.attributes.contains(.preload) ? .on : .off
            attributesMatrix.cell(withTag: ResAttributes.protected.rawValue)?.state  = resource.attributes.contains(.protected) ? .on : .off
            attributesMatrix.cell(withTag: ResAttributes.locked.rawValue)?.state     = resource.attributes.contains(.locked) ? .on : .off
            attributesMatrix.cell(withTag: ResAttributes.purgeable.rawValue)?.state  = resource.attributes.contains(.purgeable) ? .on : .off
            attributesMatrix.cell(withTag: ResAttributes.sysHeap.rawValue)?.state    = resource.attributes.contains(.sysHeap) ? .on : .off
            
            rType.stringValue = resource.type
            rID.integerValue = resource.id
            rSize.integerValue = resource.data.count
            
            // swap box
            placeholderView.contentView = resourceView
        } else if let document = currentDocument {
            // get sizes of forks as they are on disk
            dataSize.integerValue = 0
            rsrcSize.integerValue = 0
            
            // set info window elements to correct values
            self.window?.title = NSLocalizedString("Document Info", comment: "")
            if let url = document.fileURL {
                iconView.image = NSWorkspace.shared.icon(forFile: url.path)
                nameView.stringValue = url.lastPathComponent
                do {
                    let values = try url.resourceValues(forKeys: [.fileSizeKey, .totalFileSizeKey])
                    dataSize.integerValue = values.fileSize!
                    rsrcSize.integerValue = values.totalFileSize! - values.fileSize!
                } catch {}
            } else {
                iconView.image = NSImage(named: "Resource file")
                nameView.stringValue = document.displayName
            }
            
            creator.stringValue = document.hfsCreator.stringValue
            type.stringValue = document.hfsType.stringValue
            
            // swap box
            placeholderView.contentView = documentView
        }
    }
    
    private func setMainWindow(_ mainWindow: NSWindow?) {
        if let document = mainWindow?.windowController?.document as? ResourceDocument {
            currentDocument = document
            selectedResource = document.outlineView.selectedItem as? Resource
        } else {
            currentDocument = nil
            selectedResource = (mainWindow?.windowController as? ResKnifePlugin)?.resource
        }
        self.update()
    }
    
    @objc func mainWindowChanged(_ notification: Notification) {
        self.setMainWindow(notification.object as? NSWindow)
    }
    
    @objc func selectedResourceChanged(_ notification: Notification) {
        let outline = notification.object as! NSOutlineView
        if outline.window?.windowController?.document === currentDocument {
            selectedResource = outline.selectedItem as? Resource
            self.update()
        }
    }

    @objc func propertiesChanged(_ notification: Notification) {
        self.update()
    }
    
    // Check for conflicts
    func control(_ control: NSControl, isValidObject obj: Any?) -> Bool {
        guard selectedResource != nil else {
            return false
        }
        let textField = control as! NSTextField
        switch textField.identifier?.rawValue {
        case "rType":
            return selectedResource.canSetType(obj as! String)
        case "rID":
            return selectedResource.canSetID(obj as! Int)
        default:
            break
        }
        return true
    }
    
    func controlTextDidEndEditing(_ obj: Notification) {
        guard selectedResource != nil else {
            return
        }
        let textField = obj.object as! NSTextField
        switch textField.identifier?.rawValue {
        case "name":
            selectedResource.name = textField.stringValue
        case "rType":
            selectedResource.type = textField.stringValue
        case "rID":
            selectedResource.id = textField.integerValue
        case "creator":
            currentDocument.hfsCreator = OSType(textField.stringValue)
        case "type":
            currentDocument.hfsType = OSType(textField.stringValue)
        default:
            break
        }
    }
    
    @IBAction func attributesChanged(_ sender: NSButton) {
        let att = ResAttributes(rawValue: sender.selectedTag())
        selectedResource.attributes.formSymmetricDifference(att)
    }
}
